mm/context_label.py

```diff
@@ -0,0 +1,203 @@
+from typing import Tuple, Optional, Dict
+import itertools
+from dataclasses import dataclass
+
+import numpy as np
+
+from i6_core.lib.hdf import get_returnn_simple_hdf_writer
+from i6_core.lib.rasr_cache import FileArchive
+from sisyphus import Job, Task, tk
+
+
+@dataclass
+class DenseLabelInfo:
+    """
+    Attributes:
+        n_contexts: number of phonemes in lexicon ( usually need to + 1 for non-context # in rasr)
+        use_word_end_classes: if word end class is used for no tying dense label
+        use_boundary_classes: if bounary class is used for no tying dense label
+        num_hmm_states_per_phon: the number of hmm states per phoneme
+    """
+
+    n_contexts: int
+    use_word_end_classes: bool
+    use_boundary_classes: bool
+    num_hmm_states_per_phon: int
+
+
+class GetPhonemeLabelsFromNoTyingDense(Job):
+    def __init__(
+        self,
+        alignment_cache_path: tk.Path,
+        allophone_path: tk.Path,
+        dense_tying_path: tk.Path,
+        dense_label_info: DenseLabelInfo,
+        sparse: bool = False,
+        returnn_root: Optional[tk.Path] = None,
+    ):
+        """
+        Get past/center/future context label of alignment by calculating back labels from dense tying and write the
+        labels into hdf file.
+        (C.f. NoStateTyingDense in rasr
+        https://github.com/rwth-i6/rasr/blob/a942e3940c30eeba900c873f3bfb3f48d5b39ddb/src/Am/ClassicStateTying.cc#L272)
+
+        :param alignment_cache_path: path to alginment cache
+        :param allophone_path: path to allohone
+        :param dense_tying_path: path to denser tying file
+        :param dense_label_info: the dense label information
+        :param sparse: writes the data to hdf in sparse format
+        :param returnn_root: path to returnn root
+        """
+        self.alignment_cache_path = alignment_cache_path
+        self.allophone_path = allophone_path
+        self.dense_tying_path = dense_tying_path
+        assert not (
+            dense_label_info.use_boundary_classes and dense_label_info.use_word_end_classes
+        ), "we do not use both class distinctions"
+        self.dense_label_info = dense_label_info
+        self.sparse = sparse
+        self.returnn_root = returnn_root
+
+        self.out_hdf_left_context = self.output_path("left_context.hdf")
+        self.out_hdf_right_context = self.output_path("right_context.hdf")
+        self.out_hdf_center_context = self.output_path("center_context.hdf")
+
+        self.rqmt = {"cpu": 1, "mem": 8, "time": 0.5}
+
+    def tasks(self):
+        yield Task("run", resume="run", rqmt=self.rqmt)
+
+    @classmethod
+    def get_tying(cls, dense_tying_path: tk.Path) -> Tuple[Dict, int]:
```

Comment:
I would also add some documentation what exactly the returned dict is representing. I guess "state repr -> state idx" or sth like that?